bigchaindb/core.py

```diff
@@ -134,9 +135,13 @@ def end_block(self, request_end_block):
 
         # TODO: calculate if an election has concluded
         # NOTE: ensure the local validator set is updated
-        # validator_updates = self.bigchaindb.get_validator_update()
-        # validator_updates = [encode_validator(v) for v in validator_updates]
-        validator_updates = []
+        validator_updates = self.bigchaindb.get_validator_update(self.block_transactions)
```

Comment:
The logic of figuring out the new validator set is currently spread across `BigchainDB.get_validator_update`, `new_validator_set`, and `ValidatorElection.conclude`, what makes it hard to reason about how to implement new types of validator set updates in the future.

What do you think about the following alternative?

```
validator_set = get_validator_set()
new_validator_set = ValidatorElection.get_new_validator_set(txs, validator_set)
store_validator_set(new_validator_set)
```
This way, all the details of deciding how the new validator set should look like based on the block are hidden inside the method specific to this particular election process.

It is then extendable in a way that multiple processes can be stacked up:

```
yet_new_validator_set = RemoveValidatorsElection.get_new_validator_set(txs, new_validator_set)
store_validator_set(yet_new_validator_set)
```

Reply:
> The logic of figuring out the new validator set is currently spread across BigchainDB.get_validator_update, new_validator_set

I agree. But the approach you suggested has some issues i.e. `ValidatorElection.get_new_validator_set(txs, validator_set)` internally would look as follows (in `upsert_validator/validator_election.py`),

```python
        votes = {}
        for txn in txns:
            if isinstance(txn, ValidatorElectionVote):
                election_id = txn.asset['id']
                election_votes = votes.get(election_id, [])
                votes[election_id] = election_votes.append(txn)

                election = ValidatorElection.conclude(self, election_id, election_votes)
                # Once an election concludes any other conclusion for the same
                # or any other election is invalidated
                if election:
                    return [election.asset['data']]
        return []
```

The issue is that I cannot use `ValidatorElectionVote` because of import issues. The only way out I see is to create a base election class?

Reply:
What if we make `get_new_validator_set` a method of `ValidatorElectionVote` instead? Then it should be fine, right?

Reply:
> election = ValidatorElection.conclude(self, election_id, election_votes)

Tried it, could not include `ValidatorElection` class there because of import error.